src/bnfgen_cli.ml

```diff
@@ -66,13 +69,16 @@ let () =
   match g with
   | Error msg -> Printf.eprintf "Could not load grammar from %s.\n%s%!\n" !filename msg; exit 1
   | Ok g ->
-    begin match !action with
-    | Dump -> Printf.printf "%s\n" @@ Bnfgen.grammar_to_string g
-    | Reduce ->
-      let out_fun = if !buffering then print_string else (Printf.printf "%s%!") in
-      let res = Bnfgen.generate ~settings:!settings out_fun g !start_symbol in
-      begin match res with
-      | Ok _ -> print_string "\n"
-      | Error msg -> Printf.eprintf "%s%!\n" msg
+    for sentence = 1 to !sentences do
+      if !settings.debug then Printf.ksprintf !settings.debug_fun "Outputting Sentence %d of %d%!" sentence !sentences;
+      begin match !action with
+      | Dump -> Printf.printf "%s\n" @@ Bnfgen.grammar_to_string g
```

Comment:
`--dump` does not produce anything, it just outputs the grammar in its "canonicalized" form . The logic for that should be outside of the loop — the loop should be inside the `Reduce` branch. No one in their right mind will want to see the same rules more than one time. ;)

Reply:
Makes sense